### PR TITLE
Renew subscription feature

### PR DESCRIPTION
**Pull request by PinkCode15**
This PR adds the implementation for renew subscription feature

### CODE REVIEW

File: database/migrations/2024_05_24_005212_add_foreign_keys_to_transactions.php [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+{
+    /**
+     * Run the migrations.
+     *
+     * @return void
+     */
+    public function up()
+    {
+        Schema::table('transactions', function (Blueprint $table) {
+            $table->foreign('user_id')->references('id')->on('users');
```
Comment (IDTitanium) [inline database/migrations/2024_05_24_005212_add_foreign_keys_to_transactions.php:17]: Do these columns already exist and you are just adding FKs to them?

File: app/Models/Subscription.php [hunk 2]
```diff
@@ -83,18 +83,39 @@ public function getUserUidAttribute()
...
     public static function createNew(array $data): self | null
     {
+        $categoryId = null;
+        
+        if(isset($data['parent_id'])){
+            $parentSubscription = Subscription::where('id', $data['parent_id'])->first();
+            $data['name'] = $parentSubscription->name;
+            $data['url'] = $parentSubscription->url;
+            $categoryId = $parentSubscription->category_id;
+        }
```
Comment (IDTitanium) [inline app/Models/Subscription.php:109]: This method is getting too large, let's move the parent logic to another method.

File: database/migrations/2024_05_24_005212_add_foreign_keys_to_transactions.php (hunk 1 above)
Comment (PinkCode15) [reply database/migrations/2024_05_24_005212_add_foreign_keys_to_transactions.php:17]: Yes it does

File: app/Models/Subscription.php (hunk 2 above)
Comment (PinkCode15) [reply app/Models/Subscription.php:109]: Okay

File: app/Actions/Subscription/RenewSubscriptionAction.php [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+class RenewSubscriptionAction
+{
+    use FormatApiResponse;
+
+   /**
+    * @param string $parentId
+    * @param array $data
+    * @return JsonResponse
+    */
+    public function execute($parentId, array $data): JsonResponse
```
Comment (IDTitanium) [inline app/Actions/Subscription/RenewSubscriptionAction.php:19]: Include the data type for $parentId

File: app/Actions/Subscription/RenewSubscriptionAction.php [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+            $data['user_id'] = $user->id;
+            $data['currency_id'] = Currency::where('code', $data['currency'])->first()->id;
+            $data['parent_id'] = $parentSubscription->id;
+            $data['name'] = $parentSubscription->name;
+            $data['url'] = $parentSubscription->url;
+            $data['category_id'] = $parentSubscription->category_id;
+    
+            $subscription = Subscription::createNew($data);
+    
+            if(!$subscription){
```
Comment (IDTitanium) [inline app/Actions/Subscription/RenewSubscriptionAction.php:39]: Wouldn't this clash with the current existing subscription you are trying to renew? All those details will already exist.

Comment (PinkCode15) [reply app/Actions/Subscription/RenewSubscriptionAction.php:39]: No, the start and end date would be different. Also the currency and description might be different as well

Comment (IDTitanium) [reply app/Actions/Subscription/RenewSubscriptionAction.php:39]: Currency will not be different for 99% of cases, description is optional. 
however, it does make sense that start and end date will be different.

**Review by IDTitanium — APPROVED** [review]

**Pull request merged**